proposals/windows-mdm-commands-tables.md

```diff
@@ -0,0 +1,118 @@
+for Windows has two tables, with the following descriptions:
+
+### `windows_mdm_pending_commands`
+
+> The table windows_mdm_pending_commands contains the stateless commands enqueued into the MDM pending commands table from fleetctl or fleet Server.
+> These commands contain operations to be performed on a given device. 
+> They don't yet have the MessageID, SessionID, and CmdId information as this data gets populated during the response 
+> to a given MDM management request.
+
+### `windows_mdm_commands`
+
+> The windows_mdm_commands table represents the stateful commands that were actually sent over the wire. 
+> These commands include the MessageID, SessionID, and CmdId as they were sent to the device. 
+> They are stored in the commands table for error code tracking purposes. 
+> Additionally, this table can contain commands that were not part of the windows_mdm_pending_commands table and 
+> were generated by the Fleet server in response to an incoming MDM command.
+
+## Proposal
+
+I think it's fundamental we have:
+
+- A clear queue model
+- A single source of truth for commands
+- The status of a command at any point in time easily available
+- The raw XML we sent to the device
+- The raw XML we received as a response
+
+A rough version of the schema (many missing fields from the original) could be:
+
+```sql
+CREATE TABLE windows_mdm_commands (
+    -- managed and generated by Fleet
+    command_uuid    VARCHAR(127) NOT NULL,
+
+    -- raw XML
+    raw_command      MEDIUMTEXT   NOT NULL,
+
+    -- command_id specifies a unique command identifier for the SyncML message (Section 2.2.3.2 on MS-MDM spec)
+    command_id      VARCHAR(255) NOT NULL,
+
+    -- cmd_verb is e.g. Add, Exec, Get, etc.    
+    cmd_verb        VARCHAR(20)  NOT NULL,
+
+    created_at      TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_at      TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+
+    PRIMARY KEY (command_uuid, device_id)
+)
+
+CREATE TABLE windows_mdm_command_queue (
+    device_id    VARCHAR(255) NOT NULL,
+    command_uuid VARCHAR(127) NOT NULL,
+
+    active   BOOLEAN NOT NULL DEFAULT 1,
+    priority TINYINT NOT NULL DEFAULT 0,
+
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+
+    PRIMARY KEY (device_id, command_uuid),
+
+    INDEX (priority DESC, created_at),
+
+    FOREIGN KEY (device_id)
+        REFERENCES mdm_windows_enrollments (id)
+        ON DELETE CASCADE ON UPDATE CASCADE,
+
+    FOREIGN KEY (command_uuid)
+        REFERENCES commands (command_uuid)
+        ON DELETE CASCADE ON UPDATE CASCADE
+)
+
+CREATE TABLE windows_command_results (
+    device_id    VARCHAR(255) NOT NULL,
+    command_uuid VARCHAR(127) NOT NULL,
+    status       VARCHAR(31)  NOT NULL,
+
+    -- raw response, as received from the client
+    result       MEDIUMTEXT   NOT NULL,
```

Comment:
@roperzh there are two results to track here

The full SyncML message response (which contains the header + body). The command profile feature requires tracking the entire message, as this is what is shown on the CLI. This will contain not only the response for the given protocol command, but for other protocol commands sent together on the same SyncML message. I discussed this with @noahtalerman and he wanted to show the entire raw SyncML response on the CLI. 

Besides this, we need to track the result for the given protocol command. This is going to arrive as protocol command response. So I would suggest that we add another field for this.

```
    -- raw SyncML message response, as received from the client. 
    -- entire response SyncML message is stored here
    full_result       MEDIUMTEXT   NOT NULL,
    
    -- raw SyncML protocol command response, as received from the client.
    -- Result protocol command contains responses for a given command
    result       MEDIUMTEXT   NOT NULL,
```

Example protocol command request sent from the server to the device

```
     <Get>
       <CmdID>5</CmdID>
       <Item>
         <Target>
           <LocURI>./DevDetail/Ext/Microsoft/DeviceName</LocURI>
         </Target>
       </Item>
     </Get>
```

Example protocol command response sent from the device to the server

```
     <Results>
       <CmdID>3</CmdID>
       <MsgRef>1</MsgRef>
       <CmdRef>5</CmdRef>
       <Item>
         <Source>
           <LocURI>./DevDetail/Ext/Microsoft/DeviceName</LocURI>
         </Source>
         <Data>DESKTOP-DC2BRCA</Data>
       </Item>
     </Results>
```

Reply:
The issue I see is that SyncML message response (`XML3` in the previous comment) could have things unrelated to a given protocol command (`XML1`). If you check the figma [here](https://www.figma.com/file/eC1C8kp9grdPMAZGvR0ibT/%2313281-Enforce-configuration-on-Windows?type=design&node-id=240-5662&mode=design&t=HoReIfenf2uIxYVe-0), this would mean that for every protocol command sent, there would be full message responses shown on the CLI, which might not be ideal

On the other hand, just showing `XML2` would contain that raw XML of the protocol command response. 

We could go ahead and keep track of both in the DB. Then, CLI will show `XML3` by default. Then, we make this configurable through an option so the admin can decide only to see `XML2` if needed.

Reply:
@marcosd4h I'm confused, you mentioned that Noah (and the figma shows that too) wants the full SyncML to be displayed in the CLI as the result for a command, with other unrelated command results too.

> I discussed this with noahtalerman and he wanted to show the entire raw SyncML response on the CLI.

If so, it means we only need `XML3` to be saved in the DB? I mean, it seems to me that showing `XML2` would be nicer/straight to the point, but I assume we have good reasons to display `XML3` instead (or is it not set in stone yet?), and if so, we have no need for the `XML2` to be stored?

Reply:
(that being said, not a big deal either way, we can store both, just that I think `XML3` should be stored only once and the commands it contains would have a reference to its row).

Reply:
Hey @marcosd4h and @mna can you please schedule 30 mins to chat about this w/ me? 

I think I don't full grasp what the Fleet server gets back from the device (response). Is it two separate responses?`XML2` and `XML3. Or, is it just one response? `XML3`

I think we only need to store `XML3` because that's all we want to display but I want to understand how this approach affects the custom commands and profiles feature before we decide.

Reply:
> Should we have a different table to story it only once?

@marcosd4h yeah that's what I'd recommend.